encryption/src/main/java/org/apache/solr/encryption/EncryptionRequestHandler.java

```diff
@@ -260,12 +330,116 @@ public void handleRequestBody(SolrQueryRequest req, SolrQueryResponse rsp) throw
+      }
+    } finally {
+      if (success) {
+        rsp.add(STATUS, STATUS_SUCCESS);
+      } else {
+        rsp.add(STATUS, STATUS_FAILURE);
+      }
+      if (collectionState != null) {
+        rsp.add(ENCRYPTION_STATE, collectionState.value);
+      }
+      if (log.isInfoEnabled()) {
+        long timeMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeNs);
+        log.info("Responding encryption distributed state={} success={} for keyId={} collection={} timeMs={}",
+            (collectionState == null ? null : collectionState.value), success, keyId, collectionName, timeMs);
+      }
+    }
+  }
+
+  private SolrClientHolder getHttpSolrClient(SolrQueryRequest req) {
+    CoreContainer coreContainer = req.getCore().getCoreContainer();
+    CloudSolrClient cloudSolrClient = coreContainer.getSolrClientCache().getCloudSolrClient(coreContainer.getZkController().getZkClient().getZkServerAddress());
+    if (cloudSolrClient instanceof CloudHttp2SolrClient) {
+      return new SolrClientHolder(((CloudHttp2SolrClient) cloudSolrClient).getHttpClient(), false);
+    }
+    return new SolrClientHolder(new Http2SolrClient.Builder().build(), true);
+  }
+
+  protected ModifiableSolrParams createDistributedRequestParams(SolrQueryRequest req, SolrQueryResponse rsp, String keyId) {
+    ModifiableSolrParams params = new ModifiableSolrParams();
+    params.set(PARAM_KEY_ID, keyId == null ? NO_KEY_ID : keyId);
+    return params;
+  }
+
+  boolean isTimeout(long maxTimeNs) {
+    return System.nanoTime() > maxTimeNs;
+  }
+
+  private State sendEncryptionRequestWithRetry(Replica replica, ModifiableSolrParams params, Http2SolrClient httpSolrClient, String keyId, String collection) {
+    for (int numAttempts = 0; numAttempts < DISTRIBUTION_MAX_ATTEMPTS; numAttempts++) {
+      try {
+        NamedList<Object> response = sendEncryptionRequest(replica, params, httpSolrClient);
+        Object responseStatus = response.get(STATUS);
+        Object responseState = response.get(ENCRYPTION_STATE);
+        log.info("Encryption state {} status {} for replica {} keyId {} collection={}", responseStatus, responseState, replica.getName(), keyId, collection);
+        if (responseState != null) {
+          return State.fromValue(responseState.toString());
+        }
+      } catch (SolrServerException | IOException e) {
+        log.error("Error occurred while sending encryption request", e);
+      }
+    }
+    throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "Failed encryption request to replica " + replica.getName() + " for keyId " + keyId + " collection " + collection);
+  }
+
+  private NamedList<Object> sendEncryptionRequest(Replica replica, ModifiableSolrParams params, Http2SolrClient httpSolrClient)
+      throws SolrServerException, IOException {
+    GenericSolrRequest request = new GenericSolrRequest(SolrRequest.METHOD.POST, getPath(), params);
+    request.setBasePath(replica.getCoreUrl());
```

Comment:
setBasePath is now gone in Solr main.  Instead, you should use `httpSolrClient.requestWithBaseUrl`

Reply:
Finally, Solr 9.8.0 was released during this long lasting PR, so I updated.